pdns/dnsdistdist/dnsdist-backend.cc

```diff
@@ -212,6 +219,197 @@ void DownstreamState::incCurrentConnectionsCount()
+bool DownstreamState::s_randomizeIDs{false};
+int DownstreamState::s_udpTimeout{2};
+
+static bool isIDSExpired(IDState& ids)
+{
+  auto age = ids.age++;
+  return age > DownstreamState::s_udpTimeout;
+}
+
+void DownstreamState::handleTimeout(IDState& ids)
+{
+  /* We mark the state as unused as soon as possible
+     to limit the risk of racing with the
+     responder thread.
+  */
+  auto oldDU = ids.du;
+
+  ids.du = nullptr;
+  handleDOHTimeout(DOHUnitUniquePtr(oldDU, DOHUnit::release));
+  oldDU = nullptr;
+  ids.age = 0;
+  reuseds++;
+  --outstanding;
+  ++g_stats.downstreamTimeouts; // this is an 'actively' discovered timeout
+  vinfolog("Had a downstream timeout from %s (%s) for query for %s|%s from %s",
+           remote.toStringWithPort(), getName(),
+           ids.qname.toLogString(), QType(ids.qtype).toString(), ids.origRemote.toStringWithPort());
+
+  struct timespec ts;
+  gettime(&ts);
+
+  struct dnsheader fake;
+  memset(&fake, 0, sizeof(fake));
+  fake.id = ids.origID;
+
+  g_rings.insertResponse(ts, ids.origRemote, ids.qname, ids.qtype, std::numeric_limits<unsigned int>::max(), 0, fake, remote, getProtocol());
+}
+
+void DownstreamState::handleTimeouts()
+{
+  if (s_randomizeIDs) {
+    auto map = d_idStatesMap.lock();
+    for (auto it = map->begin(); it != map->end(); ) {
+      auto& ids = it->second;
+      if (isIDSExpired(ids)) {
+        handleTimeout(ids);
+        it = map->erase(it);
+        continue;
+      }
+      ++it;
+    }
+  }
+  else {
+    for (IDState& ids : idStates) {
+      int64_t usageIndicator = ids.usageIndicator;
+      if (IDState::isInUse(usageIndicator) && isIDSExpired(ids)) {
+        if (!ids.tryMarkUnused(usageIndicator)) {
+          /* this state has been altered in the meantime,
+             don't go anywhere near it */
+          continue;
+        }
+
+        handleTimeout(ids);
+      }
+    }
+  }
+}
+
+IDState* DownstreamState::getExistingState(unsigned int stateId)
+{
+  if (s_randomizeIDs) {
+    auto map = d_idStatesMap.lock();
+    auto it = map->find(stateId);
+    if (it == map->end()) {
+      return nullptr;
+    }
+    return &it->second;
+  }
+  else {
+    if (stateId >= idStates.size()) {
+      return nullptr;
+    }
+    return &idStates[stateId];
+  }
+}
+
+void DownstreamState::releaseState(unsigned int stateId)
+{
+  if (s_randomizeIDs) {
+    auto map = d_idStatesMap.lock();
+    auto it = map->find(stateId);
+    if (it == map->end()) {
+      return;
+    }
+    if (it->second.isInUse()) {
+      return;
+    }
+    map->erase(it);
+  }
+}
+
+IDState* DownstreamState::getIDState(unsigned int& selectedID, int64_t& generation)
+{
+  DOHUnitUniquePtr du(nullptr, DOHUnit::release);
+  IDState* ids = nullptr;
+  if (s_randomizeIDs) {
+    /* if the state is already in use we will retry,
+       up to 5 five times. The last selected one is used
+       even if it was already in use */
+    size_t remainingAttempts = 5;
```

Comment:
If the state table is quite full and/or if you have bad luck this will start reusing IDs. What are the consequences of that? Should that be documented?

Reply:
That already happens in master since we have a fixed number of possible states for in-flight queries, and is tracked in our stats as `reused`. Basically it means that your query will appear to have had a time out much faster than expected. I do not think it is a real issue for the use case this feature tries to address.